File: tests/patterns/python/aws_dynamodb_to_firestore/before.py
import boto3

dynamodb = boto3.resource('dynamodb')
table = dynamodb.Table('users')


def create_user(user_id, name, email):
    table.put_item(
        Item={
            'user_id': user_id,
            'name': name,
            'email': email,
        }
    )


def get_user(user_id):
    response = table.get_item(
        Key={'user_id': user_id}
    )
    return response.get('Item')


def update_user(user_id, name):
    table.update_item(
        Key={'user_id': user_id},
        UpdateExpression='SET #n = :name',
        ExpressionAttributeNames={'#n': 'name'},
        ExpressionAttributeValues={':name': name},
    )


def delete_user(user_id):
    table.delete_item(
        Key={'user_id': user_id}
    )


def query_users_by_email(email):
    response = table.query(
        IndexName='email-index',
        KeyConditionExpression='email = :email',
        ExpressionAttributeValues={':email': email},
    )
    return response['Items']

File: tests/patterns/typescript/aws_dynamodb_to_firestore/before.ts
import {
  DynamoDBClient,
  PutItemCommand,
  GetItemCommand,
  DeleteItemCommand,
  QueryCommand,
} from "@aws-sdk/client-dynamodb";
import { marshall, unmarshall } from "@aws-sdk/util-dynamodb";

const client = new DynamoDBClient({ region: "us-east-1" });
const TABLE_NAME = "users";

interface User {
  userId: string;
  name: string;
  email: string;
}

export async function createUser(user: User): Promise<void> {
  const command = new PutItemCommand({
    TableName: TABLE_NAME,
    Item: marshall(user),
  });
  await client.send(command);
}

export async function getUser(userId: string): Promise<User | null> {
  const command = new GetItemCommand({
    TableName: TABLE_NAME,
    Key: marshall({ userId }),
  });
  const response = await client.send(command);
  if (!response.Item) return null;
  return unmarshall(response.Item) as User;
}

export async function deleteUser(userId: string): Promise<void> {
  const command = new DeleteItemCommand({
    TableName: TABLE_NAME,
    Key: marshall({ userId }),
  });
  await client.send(command);
}

export async function queryUsersByEmail(email: string): Promise<User[]> {
  const command = new QueryCommand({
    TableName: TABLE_NAME,
    IndexName: "email-index",
    KeyConditionExpression: "email = :email",
    ExpressionAttributeValues: marshall({ ":email": email }),
  });
  const response = await client.send(command);
  return (response.Items ?? []).map((item) => unmarshall(item) as User);
}
